Source/gui/SNTBinaryMessageWindowController.m

```diff
@@ -135,13 +150,96 @@ - (IBAction)showCertInfo:(id)sender {
+    // TODO: handle error
+    LOGE(@"Unable to process Touch ID error: %@", err);
+    return;
+  }
+
+  [context evaluatePolicy:LAPolicyDeviceOwnerAuthenticationWithBiometrics
+          localizedReason:[NSString stringWithFormat:@"Approve %@", self.event.signingID]
+                    reply:^(BOOL success, NSError *error) {
+                      if (success) {
+                        SNTRuleType ruleType = SNTRuleTypeSigningID;
+                        NSString *ruleIdentifier = self.event.signingID;
+
+                        // Check here to see if the binary is validly signed if not
+                        // then use a hash rule instead of a signing ID
+                        if (self.event.signingChain.count == 0) {
+                          LOGD(@"No certificate chain found for %@", self.event.filePath);
+                          ruleType = SNTRuleTypeBinary;
+                          ruleIdentifier = self.event.fileSHA256;
+                        }
+
+                        // Add rule to allow binary same as santactl rule.
+                        SNTRule *newRule =
+                          [[SNTRule alloc] initWithIdentifier:ruleIdentifier
+                                                        state:SNTRuleStateAllow
+                                                         type:ruleType
+                                                    customMsg:@"Approved by user"
+                                                    timestamp:[[NSDate now] timeIntervalSince1970]];
+
+                        // TODO add bundle hash to a dictionary / database
+                        // table for auto-approval so the user isn't drowning
+                        // in dialogs
+                        dispatch_semaphore_t sema2 = dispatch_semaphore_create(0);
+
+                        MOLXPCConnection *daemonConn =
+                          [SNTXPCControlInterface configuredConnection];
+                        daemonConn.invalidationHandler = ^{
+                          LOGE(@"Connection invalidated");
+                          dispatch_semaphore_signal(sema2);
+                        };
+
+                        [daemonConn resume];
+                        [[daemonConn remoteObjectProxy]
+                          addStandaloneRule:newRule
+                                      reply:^(NSError *error) {
+                                        if (error) {
+                                          LOGE(@"Failed to modify rules standalone: %s",
+                                               [error.localizedDescription UTF8String]);
+                                          LOGE(@"Failure reason: %@", error.localizedFailureReason);
+                                        }
+                                        dispatch_semaphore_signal(sema2);
+                                      }];
+                        dispatch_semaphore_wait(sema2, DISPATCH_TIME_FOREVER);
```

Comment:
We should add some timeouts.

Reply:
Good call, should this be configurable?

Reply:
Most other XPC calls we use a 5 sec timeout. That should work fine here too.

Reply:
I removed the semaphore since it's unnecessary now.